### PR TITLE
Add dispatch server to handle multiple games

### PR DESCRIPTION
**Pull request by swcraig**
This is what will be hosted online that clients will subscribe to. Anytime the board state is changed the game server will hit this dispatch server with a JSON represenation of the board state, which will be forwarded to the subscribed clients (such as the VR team).

The game servers and websockets must use a game_id argument in the url. Game servers are expected to generate unique game IDs for each game. URL example: dispatchserverdomain.ca/gidj34s

Game servers will send POST requests, WebSocket clients will send GET requests.

/list_games will send the list of currently running game IDs.

### CODE REVIEW

File: dispatch_server.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+# Responsible for relaying all board state changes to subscribed clients
+class DispatchServer < Sinatra::Base
+  set :clients, Hash.new([])
+  set :current_board_state, {}
+
+  # The game server will hit this with updates to the board
+  post '/:game_id' do
+    game_id = params[:game_id]
+    settings.current_board_state[game_id] = request.body.read
```
Comment (marclave) [inline dispatch_server.rb]: this will break if `params[:game_id]` hash does not exist in the game list of all hashes.

File: dispatch_server.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    settings.clients[game_id].each do |client|
+      client.send(settings.current_board_state[game_id])
+    end
+  end
+
+  get '/list_games' do
+    if !request.websocket?
```
Comment (marclave) [inline dispatch_server.rb]: Can we not default to `json`? Is there a reason why we have it as a string?

File: dispatch_server.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      settings.clients.keys.to_json
+    end
+  end
+
+  # Clients will open WebSocket connections to watch the game here
+  get '/:game_id?' do
+    if !request.websocket?
+      'You are not a websocket'
+    else
+      game_id = params[:game_id]
```
Comment (marclave) [inline dispatch_server.rb]: Also need to check the hash here.

File: dispatch_server.rb (hunk 1 above)
Comment (swcraig) [inline dispatch_server.rb]: It will just create a new board state representation with that "game_id" (whatever string it is).

File: Gemfile [hunk 4]
```diff
@@ -3,7 +3,10 @@ source 'https://rubygems.org'
 gem 'colorize', '0.8.1'
 gem 'factory_girl', '~> 4.0'
 gem 'pry', '0.10.4'
+gem 'rack'
 gem 'rake', '12.0.0'
 gem 'rspec', '3.5.0'
 gem 'rubocop', '0.48.1'
 gem 'simplecov', '0.14.1'
+gem 'sinatra'
```
Comment (jacquesporveau) [inline Gemfile]: Nice choice.

File: dispatch_server.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    settings.current_board_state.keys.to_json
+  end
+
+  # Clients will open WebSocket connections to watch the game here
+  get '/:game_id' do
+    if !request.websocket?
+      NOT_A_WEBSOCKET
+    else
+      game_id = params[:game_id]
+      request.websocket do |ws|
```
Comment (jacquesporveau) [inline dispatch_server.rb]: Can we rename ws to web_socket?

File: lib/move_type.rb [hunk 6]
```diff
@@ -7,30 +7,20 @@ module MoveType
   class << self
```
Comment (jacquesporveau) [inline lib/move_type.rb]: What does this file and the rest of the files below have to do with adding a dispatch server?

**Pull request merged**